docs/preview/counterExpression.md

```diff
@@ -0,0 +1,57 @@
+# Counter Expression
+
+## Overview
+
+This feature adds a new expression function, `counter`, which will atomically increment a variable value at the start of each build. The value of this variable will remain the same across phases, regardless of how or where those phases are executed.
+
+`counter` values are identified by their "prefix" which acts as a lookup key, meaning the same `counter` can be used in each phase or as a top-level variable, and it will be incremented exactly once for the build. The value of `counter` is unique to the build definition in which it is defined, regardless of the prefix.
+
+## Syntax
+
+The UI will provide an editor, but the functionality will be stored as a [variable expression](conditions.md). Users may also enter the expression directly into the current variable field.
+
+*TBD: we should discuss what the user sees when they return to the Variables tab later. Does the UI parse out `counter` variables and only show them in the new UI? Does it show up in both places?*
+
+Example: Produce a variable with a prefix from the `$BuildNumberPrefix` which starts counting from `$BuildNumberSeed`
+\
+`counter(variables['BuildNumberPrefix'], variables['BuildNumberSeed'])`
+
+Example: On the first build, the variable will have the string value `100` which then increments on each subsequent build (`101`, `102`, ...)
+\
+`counter('', 100)`
```

Comment:
are you planning to support an overload `counter('1.')` where the seed is optional? what would the first value be? `1.1`?

Reply:
or even an additional overload `counter()` which is equivalent to `counter('')`

Reply:
I wasn't planning on it, but I'm sure that can be worked out. Both overloads should be fairly trivial to parse.

From the original design I was given, counters start at `0` by default, so `counter('1.')` would result in `1.0`, `1.1`, …